vyper/venom/parser.py

```diff
@@ -0,0 +1,143 @@
+from vyper.venom.context import IRContext
+from vyper.venom.basicblock import (
+    IRLabel,
+    IRVariable,
+    IRLiteral,
+    IROperand,
+    IRInstruction,
+    IRBasicBlock,
+)
+from vyper.venom.function import IRFunction
+from lark import Lark, Transformer
+
+
+VENOM_PARSER = Lark(
+    """
+    %import common.CNAME
+    %import common.DIGIT
+    %import common.LETTER
+    %import common.WS
+    %import common.INT
+
+    start: function* data_section
+    function: "fn" NAME "=>" "{" block* "}"
+    data_section: "data:" call*
+
+    block: NAME ":" (call | assignment)*
+
+    assignment: VAR_IDENT "=" expr
+    expr: call | CONST
+    call: CNAME operands_list
+
+    operands_list: (operand ("," operand)*)?
+
+    operand: VAR_IDENT | CONST | LABEL
+    CONST: INT
+
+    VAR_IDENT: "%" INT (":" INT)?
+    LABEL: "@" NAME
+    NAME: (DIGIT|LETTER|"_")+
+
+    %ignore WS
+    """
+)
+
+
+def set_last_var(fn: IRFunction):
+    for block in fn.get_basic_blocks():
+        output_vars = (
+            instruction.output
+            for instruction in block.instructions
+            if instruction.output is not None
+        )
+        for output_var in output_vars:
+            assert isinstance(output_var, IRVariable)
+            value = output_var.value
+            assert value.startswith("%")
+            value = value.replace("%", "", 1)
+            if not value.isdigit():
+                continue
+            fn.last_variable = max(fn.last_variable, int(value))
+
+
+class VenomTransformer(Transformer):
+    def start(self, children) -> IRContext:
+        ctx = IRContext()
+        funcs = children[:-1]
+        data_section = children[-1]
+        for fn_name, blocks in funcs:
+            fn = ctx.create_function(fn_name)
+            for block_name, instructions in blocks:
+                bb = IRBasicBlock(IRLabel(block_name), fn)
+
+                for instruction in instructions:
+                    bb.insert_instruction(instruction)
+
+                # Manually insert because we need to override function entry
+                fn._basic_block_dict[block_name] = bb
+
+            set_last_var(fn)
```

Comment:
since this is more or less a local helper function, let's rename to `_set_last_var`